### PR TITLE
chore: add an automatic build shared library ci

### LINKED ISSUE #116: How about adding a ci which automatic build shared library?
## Summary 
Add a CI which automatic build the shared libraries.

## Motivation
The shared libraries are referenced by other project like finschia for building. 
we need to compile shared library every commit is added if we want to apply the changes of wasmvm, but it doesn't seems to do that often. So I propose adding CI to automatically build the shared libraries.

The `cosmwasm/wasmvm` already do auto commit it. 
https://github.com/CosmWasm/wasmvm/blob/1638725b25d799f078d053391945399cb35664b1/.circleci/config.yml#L282-L322
```yaml
  deploy_to_git:
    machine:
      image: ubuntu-2004:2022.10.1
    resource_class: xlarge
    steps:
      - add_ssh_keys:
          fingerprints:
            # Custom read/write deployment key with private key stored on CircleCI
            # (see https://app.circleci.com/settings/project/github/CosmWasm/wasmvm/ssh and https://github.com/CosmWasm/wasmvm/settings/keys)
            - "31:de:e5:84:1b:12:81:94:aa:06:50:c0:cb:bd:79:f0"
      - checkout
      - run:
          name: Build shared library for Linux
          command: make release-build-linux
      - run:
          name: Build shared library for macOS
          command: make release-build-macos
      # Shared libraries for Windows (.dll) currently do not work (https://github.com/CosmWasm/wasmvm/issues/389)
      # and .dll builds are not deterministic.
      # Deactivating this step to avoid polluting the git hostory.
      #
      # - run:
      #     name: Build shared library for Windows
      #     command: make release-build-windows
      - run:
          name: Debug build results
          command: ls -l ./internal/api
      - run:
          name: Configure git user
          # This is not a GitHub user and no permissions can be configured other than "push access", which
          # we can configure for Deploy keys at https://github.com/CosmWasm/wasmvm/settings/keys
          command: |
            git config user.email "wasmvm@circleci.confio.example.com"
            git config user.name "Deployer"
      - run:
          name: Check-in and push new libraries
          command: |
            git status
            git add ./internal/api
            git commit --allow-empty -m '[skip ci] Built release libraries'
            git push origin $CIRCLE_BRANCH
```

## Proposal
* If new commit is added in main branch, a CI automatically detect and add new PR which compile the shared libraries and change it. -> if there is no changed in the shared libraries to build, do not this.
* if the all test of the PR are passed, do auto-merge.

### PR DESCRIPTION
**Pull request by 170210**
# Description

This PR has following changes.

- add a CI which automatic build the shared libraries.

Closes #116 

## Types of changes

- [ ] Bug fix (changes which fixes an issue)
- [ ] New feature (changes which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] ETC (build, ci, docs, perf, refactor, style, test)

## Checklist

- [x] I followed the [contributing guidelines](https://github.com/Finschia/wasmvm/blob/main/CONTRIBUTING.md).
- [ ] I have updated the documentation accordingly.(not need)
- [ ] I have added tests to cover my changes.(not need)
- [x] The PR title and commits are followed [conventional commit form](https://www.conventionalcommits.org/en/v1.0.0).

### CODE REVIEW

File: .github/workflows/deploy_to_git.yml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    needs: build_shared_library
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Get app token
+        uses: tibdex/github-app-token@v1
+        id: generate-token
+        with:
+          app_id: ${{ secrets.FINSCHIA_AUTO_PR_APP_ID }}
+          private_key: ${{ secrets.FINACHIS_AUTO_PR_APP_KEY }}
```
Comment (zemyblue) [inline .github/workflows/deploy_to_git.yml:85]: This key name need to change to `FINSCHIA_AUTO_PR_APP_PRIVATE_KEY`

File: .github/workflows/deploy_to_git.yml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        os: [linux, macos]
+        include: 
+          - os: linux
+            dockerfile_name: centos7
+            shared_library_extension: so
+          - os: macos
+            dockerfile_name: cross
+            shared_library_extension: dylib
+    runs-on: ubuntu-latest
+    # excute only after pr merged
```
Comment (zemyblue) [inline .github/workflows/deploy_to_git.yml:29]: `execute`

File: internal/api/checksums.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+7392529b31b45e5e53b98139dd5880eb82c688e90cf50c62a9cf345917d74db6  libwasmvm.aarch64.so
+d7fd7799ec6e385b3d83a8b0eac1f2a547a6f74a97175ff9a40303fe9de3c646  libwasmvm.x86_64.so
+aa206401dd8f78021428349726a2d306cd8e094df49b0b42114320ab71c4d2a7  libwasmvm.dylib
```
[the remark this answers is not in the document]
Comment (170210) [reply internal/api/checksums.txt:3]: I originally wanted to use checksum to allow developers to validate the current checksum information, but it is indeed a bit redundant. I will delete this file.

File: .github/workflows/deploy_to_git.yml [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Get app token
+        uses: tibdex/github-app-token@v1
+        id: generate-token
+        with:
+          app_id: ${{ secrets.FINSCHIA_AUTO_PR_APP_ID }}
+          private_key: ${{ secrets.FINSCHIA_AUTO_PR_APP_PRIVATE_KEY }}
+      # https://github.com/tibdex/github-app-token/issues/54#issuecomment-1410471261
```
Comment (loloicci) [inline .github/workflows/deploy_to_git.yml:86]: unify indent level to the part this comment for

File: .github/workflows/deploy_to_git.yml [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            - [ ] Bug fix (changes which fixes an issue)(not need)
+            - [ ] New feature (changes which adds functionality)(not need)
+            - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)(not need)
+            - [x] ETC (build, ci, docs, perf, refactor, style, test)
+
+            ## Checklist
+            - [x] I followed the [contributing guidelines](https://github.com/Finschia/wasmvm/blob/main/CONTRIBUTING.md).
+            - [ ] I have updated the documentation accordingly.(not need)
+            - [ ] I have added tests to cover my changes.(not need)
+            - [x] The PR title and commits are followed [conventional commit form](https://www.conventionalcommits.org/en/v1.0.0).
```
Comment (loloicci) [inline .github/workflows/deploy_to_git.yml:147]: I think this checklist is not needed

**Review by loloicci — APPROVED** [review]

**Review by zemyblue — APPROVED** [review]

**Pull request merged**